spec-tests/src/test/java/org/candlepin/spec/consumers/ConsumerCheckinSpecTest.java

```diff
@@ -0,0 +1,209 @@
+        assertThat(consumer)
+            .isNotNull()
+            .doesNotReturn(initialCheckin, ConsumerDTO::getLastCheckin);
+    }
+
+    @Test
+    public void shouldUpdateLastCheckinTimeOnAsyncHypervisorCheckin() throws Exception {
+        ApiClient adminClient = ApiClients.admin();
+        OwnerDTO owner = adminClient.owners().createOwner(Owners.random());
+        ConsumerDTO consumer = adminClient.consumers().createConsumer(Consumers.random(owner));
+        ApiClient consumerClient = ApiClients.ssl(consumer);
+        OffsetDateTime initialCheckin = consumer.getLastCheckin();
+
+        hypervisorCheckin(owner, consumerClient, StringUtil.random("name-"), StringUtil.random("id-"),
+            List.of("guest-1", "guest-2", "guest-3"), Map.of("test_fact", "fact_value"),
+            StringUtil.random("reporter-"), true);
+
+        consumer = consumerClient.consumers().getConsumer(consumer.getUuid());
+        assertThat(consumer)
+            .isNotNull()
+            .doesNotReturn(initialCheckin, ConsumerDTO::getLastCheckin);
+    }
+
+    @Test
+    public void shouldNotUpdateLastCheckinWhenFetchingReleases() throws Exception {
+        ApiClient adminClient = ApiClients.admin();
+        OwnerDTO owner = adminClient.owners().createOwner(Owners.random());
+        ConsumerDTO consumer = adminClient.consumers().createConsumer(Consumers.random(owner));
+        ApiClient consumerClient = ApiClients.ssl(consumer);
+        OffsetDateTime initialCheckin = consumer.getLastCheckin();
+
+        consumerClient.consumers().getRelease(consumer.getUuid());
+
+        consumer = consumerClient.consumers().getConsumer(consumer.getUuid());
+        assertThat(consumer)
+            .isNotNull()
+            .returns(initialCheckin, ConsumerDTO::getLastCheckin);
+    }
+
+    private HypervisorUpdateResultDTO hypervisorCheckin(OwnerDTO owner, ApiClient consumerClient,
+        String hypervisorName, String hypervisorId, List<String> guestIds, Map<String, String> facts,
+        String reporterId, boolean createMissing) throws ApiException, IOException {
+        JsonNode hostGuestMapping = getAsyncHostGuestMapping(hypervisorName, hypervisorId, guestIds, facts);
+        AsyncJobStatusDTO job = consumerClient.hypervisors().hypervisorUpdateAsync(
+            owner.getKey(), createMissing, reporterId, hostGuestMapping.toString());
+        HypervisorUpdateResultDTO resultData = null;
+        if (job != null) {
+            AsyncJobStatusDTO status = consumerClient.jobs().waitForJob(job.getId());
+            assertThatJob(status).isFinished();
+            resultData = getResultData(status);
+        }
+
+        return resultData;
+    }
+
+    private JsonNode getAsyncHostGuestMapping(String name, String id, List<String> expectedGuestIds,
+        Map<String, String> facts) {
+        ObjectMapper om = ApiClient.MAPPER;
+        ObjectNode root = om.createObjectNode();
+        ArrayNode hypervisors = root.putArray("hypervisors");
```

Comment:
The various Jackson nodes here support chaining, so this could be made a bit easier to read by rearranging certain ops and chaining calls:

```
ArrayNode guestIds = mapper.createArrayNode();
expectedGuestIds.stream()
  .map(gid -> mapper.createObjectNode().put("guestId", gid))
  .forEach(guestIds::add);

ObjectNode hypervisor = mapper.createObjectNode()
  .put("name", name)
  .set("guestIds", guestIds)
  .set("facts", mapper.valueToTree(facts))
  .putObject("hypervisorId") // context change occurs here
  .put("hypervisorId", id);

return mapper.createArrayNode()
  .add(hypervisor);
```

Reply:
Updated